In [1]:
## Data preprocessing, visualization, dataset splitting
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

## Machine Learning
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier


## Evaluation
from sklearn.metrics import classification_report

In [2]:
df_emails = pd.read_csv('emails.csv')

In [3]:
df_emails.head()

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]

In [4]:
df_emails.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5172 entries, 0 to 5171
Columns: 3002 entries, Email No. to Prediction
dtypes: int64(3001), object(1)
memory usage: 118.5+ MB


In [5]:
df_emails.describe()

the           to          ect          and          for  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000   
mean      6.640565     6.188128     5.143852     3.075599     3.124710   
std      11.745009     9.534576    14.101142     6.045970     4.680522   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       0.000000     1.000000     1.000000     0.000000     1.000000   
50%       3.000000     3.000000     1.000000     1.000000     2.000000   
75%       8.000000     7.000000     4.000000     3.000000     4.000000   
max     210.000000   132.000000   344.000000    89.000000    47.000000   

                of            a          you          hou           in  ...  \
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  ...   
mean      2.627030    55.517401     2.466551     2.024362    10.600155  ...   
std       6.229845    87.574172     4.314444     6.967878    19.281892  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000    12.000000     0.000000     0.000000     1.000000  ...   
50%       1.000000    28.000000     1.000000     0.000000     5.000000  ...   
75%       2.000000    62.250000     3.000000     1.000000    12.000000  ...   
max      77.000000  1898.000000    70.000000   167.000000   223.000000  ...   

          connevey          jay       valued          lay  infrastructure  \
count  5172.000000  5172.000000  5172.000000  5172.000000     5172.000000   
mean      0.005027     0.012568     0.010634     0.098028        0.004254   
std       0.105788     0.199682     0.116693     0.569532        0.096252   
min       0.000000     0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000     0.000000        0.000000   
50%       0.000000     0.000000     0.000000     0.000000        0.000000   
75%       0.000000     0.000000     0.000000     0.000000        0.000000   
max       4.000000     7.000000     2.000000    12.000000        3.000000   

          military     allowing           ff          dry   Prediction  
count  5172.000000  5172.000000  5172.000000  5172.000000  5172.000000  
mean      0.006574     0.004060     0.914733     0.006961     0.290023  
std       0.138908     0.072145     2.780203     0.098086     0.453817  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     1.000000     0.000000     1.000000  
max       4.000000     3.000000   114.000000     4.000000     1.000000  

[8 rows x 3001 columns]

In [19]:
df_emails.corr()

the        to       ect       and       for        of  \
the         1.000000  0.852715  0.337249  0.841200  0.784112  0.796397   
to          0.852715  1.000000  0.375480  0.825474  0.781971  0.752722   
ect         0.337249  0.375480  1.000000  0.272863  0.369777  0.178028   
and         0.841200  0.825474  0.272863  1.000000  0.751287  0.809665   
for         0.784112  0.781971  0.369777  0.751287  1.000000  0.681457   
...              ...       ...       ...       ...       ...       ...   
military    0.129466  0.091639 -0.007690  0.084147  0.067151  0.073004   
allowing    0.127019  0.120059  0.004368  0.124766  0.121057  0.108786   
ff          0.341878  0.406666  0.141460  0.400225  0.301074  0.444252   
dry         0.051021  0.071388  0.002492  0.042484  0.038126  0.026403   
Prediction -0.004421  0.055277 -0.120782  0.114364 -0.003101  0.197234   

                   a       you       hou        in  ...  connevey       jay  \
the         0.784451  0.471392  0.303621  0.845670  ...  0.008926  0.075479   
to          0.896466  0.508513  0.347993  0.881759  ...  0.013250  0.101247   
ect         0.400009  0.155783  0.974152  0.298387  ...  0.134339  0.031431   
and         0.815196  0.476764  0.235953  0.874276  ...  0.005151  0.104454   
for         0.744098  0.495852  0.329051  0.762659  ...  0.022168  0.041775   
...              ...       ...       ...       ...  ...       ...       ...   
military    0.111685  0.006498  0.005429  0.120620  ... -0.002249 -0.002979   
allowing    0.105358  0.082757 -0.000966  0.138099  ... -0.002675 -0.003543   
ff          0.464473  0.195058  0.114210  0.448303  ...  0.005403  0.073690   
dry         0.093822  0.028883  0.000601  0.077751  ... -0.003373  0.035028   
Prediction  0.107776  0.130293 -0.128340  0.154055  ... -0.030375 -0.031694   

              valued       lay  infrastructure  military  allowing        ff  \
the         0.225586  0.223426        0.101768  0.129466  0.127019  0.341878   
to          0.232847  0.255793        0.093322  0.091639  0.120059  0.406666   
ect         0.046080  0.061550        0.004393 -0.007690  0.004368  0.141460   
and         0.272963  0.253440        0.151980  0.084147  0.124766  0.400225   
for         0.236213  0.213631        0.134469  0.067151  0.121057  0.301074   
...              ...       ...             ...       ...       ...       ...   
military    0.043408  0.104297        0.041300  1.000000  0.055227  0.049524   
allowing   -0.005130  0.018550        0.276001  0.055227  1.000000  0.096212   
ff          0.130356  0.164296        0.114092  0.049524  0.096212  1.000000   
dry        -0.006468  0.018939       -0.003137  0.010835 -0.003995  0.049690   
Prediction  0.098775  0.064315        0.038161  0.064850  0.011279  0.135479   

                 dry  Prediction  
the         0.051021   -0.004421  
to          0.071388    0.055277  
ect         0.002492   -0.120782  
and         0.042484    0.114364  
for         0.038126   -0.003101  
...              ...         ...  
military    0.010835    0.064850  
allowing   -0.003995    0.011279  
ff          0.049690    0.135479  
dry         1.000000   -0.006260  
Prediction -0.006260    1.000000  

[3001 rows x 3001 columns]

In [7]:
X = df_emails.drop(['Email No.', 'Prediction'], axis = 'columns')
y = df_emails['Prediction']

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42, stratify=y)

In [30]:
## create all model for prediction
logistic_model = LogisticRegression(max_iter=1000)
knn_model = KNeighborsClassifier()
dtree_model = DecisionTreeClassifier()
svm_model = SVC()
NB_model = GaussianNB()
NN_model = MLPClassifier()

In [31]:
logistic_model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [33]:
knn_model.fit(X_train, y_train)
dtree_model.fit(X_train, y_train)
svm_model.fit(X_train, y_train)
NB_model.fit(X_train, y_train)
NN_model.fit(X_train, y_train)

MLPClassifier()

In [34]:
logi_pred = logistic_model.predict(X_test)
knn_pred = knn_model.predict(X_test)
dtree_pred = dtree_model.predict(X_test)
svmmodel_pred = svm_model.predict(X_test)
NB_model_pred = NB_model.predict(X_test)
NN_model_pred = NN_model.predict(X_test)

In [39]:
## print classification report
print("\033[1m" + "LOGISTIC REGRESSION RESULT : " + "\033[0m")
print(classification_report(logi_pred, y_test))

print("\033[1m" + "KNN RESULT : " + "\033[0m")
print(classification_report(knn_pred, y_test))

print("\033[1m" + "DECISION TREE RESULT : " + "\033[0m")
print(classification_report(dtree_pred, y_test))

print("\033[1m" + "SVM RESULT : " + "\033[0m")
print(classification_report(svmmodel_pred, y_test))

print("\033[1m" + "NAIVE BAYES RESULT : "  + "\033[0m")
print(classification_report(NB_model_pred, y_test))

print("\033[1m" + "NEURAL NETWORK RESULT : " + "\033[0m")
print(classification_report(NN_model_pred, y_test))

LOGISTIC REGRESSION RESULT : 
              precision    recall  f1-score   support

           0       0.98      0.99      0.99      1206
           1       0.97      0.96      0.97       501

    accuracy                           0.98      1707
   macro avg       0.98      0.97      0.98      1707
weighted avg       0.98      0.98      0.98      1707

KNN RESULT : 
              precision    recall  f1-score   support

           0       0.87      0.93      0.90      1136
           1       0.84      0.73      0.78       571

    accuracy                           0.86      1707
   macro avg       0.85      0.83      0.84      1707
weighted avg       0.86      0.86      0.86      1707

DECISION TREE RESULT : 
              precision    recall  f1-score   support

           0       0.95      0.95      0.95      1213
           1       0.87      0.88      0.88       494

    accuracy                           0.93      1707
   macro avg       0.91      0.91      0.91      1707
weight

Dari hasil yang didapat untuk 6 model yang dilakukan pengujian
jika diurut secara ranking, model dengan akurasi dan F1 terbaik didapatkan oleh

1) Neural Network (accuracy = 99%, F1 Score 0 : 99%, F1 Score 1 : 98%)

2) Logistic Regression (accuracy = 98%, F1 Score 0 : 99%, F1 Score 1 : 97%)

3) Naive Bayes (accuracy = 95%, F1 Score 0 : 96%, F1 Score 1 : 92%) 

4) Decision Tree (accuracy = 93%, F1 Score 0 : 95%, F1 Score 1 : 88%)

5) K-Nearest Neighbors (accuracy = 86%, F1 Score 0 : 90%, F1 Score 1 : 78%)

6) Support Vector Machine (SVM) (accuracy = 80%, F1 Score 0 : 87%, F1 Score 1 : 50%)

secara model performance (tanpa tuning parameter), hasil terbaik diraih oleh neural network, tapi melihat peringkat ke 2 dengan akurasi dan F1 score yang tidak jauh berbeda logistic regression juga merupakan pilihan yang baik dalam implementasi model dikarenakan konsep penghitungan neural network yang cukup rumit dan computational power yang tinggi untuk melakukan implementasi dari model neural network.